### ISSUE TITLE
Bring consult of GEDCOM royals down to <1000 ms.

### ISSUE THREAD

**Issue by jburse** [jburse/dogelog-moon#175]
Interestingly it is feasible to load the GEDCOM Royals from
withing the JavaScript Version of Dogelog Runtime. We find:
```
/* Dogelog Runtime 0.9.5 */
:- time(['gedcom/royal92.pl']).
% Wall 4058 ms, gc 0 ms, 1571659 lips
```
On the other hand Jekejeke Prolog does it much faster. But
the speed is higher than the usual ration between the two systems:
```
/* Jekejeke Prolog 1.5.1 */
?- consult('<path>/gedcom/royal92.pl').
% 1 consults and 0 unloads in 156 ms.
Yes
```
The reason are possibly not the sequence like grammar 
productions for Prolog terms and Prolog tokens. Rather
the grammar productions that branch into different variants.

Currently the Prolog interpreter will try all branches and
not be able to apply some indexing. This is because monadic
GCD, which is the programming technique behind the 

tokenizer and parser, is not ammenable to indexing. It would
require that the monadic predicates are inlined. And even then
it would require deep indexing. But we could manually optimize

the tokenizer and parser, in that we first du current_XXX and
then have a subpredicate with first argument indexing. Currious
whether this would give some speed.

**Comment by jburse** [comment]
To have indexing in the tokenizer, we will possibly need to 
introduce code_class/2. This would group code_type/2 values.
We such a thingy in Jekejeke Prolog, but code_class/2 was

recently removed from Jekejeke Prolog, when we introduced
code_type/2 for Jekejeke Prolog, to parallel Dogelog runtime.
But we could now parallel Dogelog runtime and 

introduce code class on both sides.

See also:

**Rename code_class/2 from library(text)** 
[https://\github.com/jburse/jekejeke-devel/issues/186](https://github.com/jburse/jekejeke-devel/issues/186)

**Comment by jburse** [comment]
We could already increase the consult speed by introducing
code_class/2. The measurement is now:
```
:- time(['gedcom/royal92.pl']).
% Wall 2766 ms, gc 3 ms, 1711908 lips
```
But the code still contains a lot of Prolog anti-patterns. For
example because of comfort we use:
```
is_name(C) :- code_class(C, D), member(D, [is_ident, is_var, is_digit]).
```
This should be replaced by a disjunction or a predicate definition,
so that indexing can do its work.
```

**Comment by jburse** [comment]
We have replaced is_name/1 by the following., calling first 
code_class/2 and then calling name_class/1:
```
name_class(is_ident).
name_class(is_var).
name_class(is_digit).
```
Interestingly this did indeed increase the speed again:
```
:- time(['gedcom/royal92.pl']).
% Wall 2536 ms, gc 0 ms, 1757406 lips
```
But possibly we have reached the end of optimizing the tokenizer
and parser. Further optimizatiions will only result from other ideas
concerning optimizing the Prolog interpreter in general.

Therefore closing the ticket.
